lib/result.rb

```diff
@@ -0,0 +1,120 @@
+# frozen_string_literal: true
+# typed: strict
+
+module Result
+  class UncheckedError < RuntimeError; end
+
+  class Base
+    def initialize(value)
+      @value = value
+    end
+
+    protected
+
+    attr_reader :value
+  end
+
+  class Ok < Base
+    def ok?
+      true
+    end
+
+    def error?
+      false
+    end
+
+    def ok_value
+      value
+    end
+
+    def ok_value_or(_fallback_value)
+      value
+    end
+
+    def ok_value_or_else(&_fallback_block)
+      value
+    end
+
+    def error_value
```

Comment:
I might have this raise an exception. Accessing the error_value when there is no error is a breaking error I think.

Reply:
Looking at the Error okay_value I would agree with Fred here.

Reply:
Happy to change this behaviour to raise. I wasn't sure about this one.